plugins/woocommerce/includes/wc-user-functions.php

```diff
@@ -424,19 +424,27 @@ function wc_customer_bought_product( $customer_email, $user_id, $product_id ) {
 	 */
 	$use_lookup_tables = apply_filters( 'woocommerce_customer_bought_product_use_lookup_tables', false, $customer_email, $user_id, $product_id );
 
-	$transient_name = 'wc_customer_bought_product_' . md5( $customer_email . $user_id . $use_lookup_tables );
-
 	if ( $use_lookup_tables ) {
 		// Lookup tables get refreshed along with the `woocommerce_reports` transient version.
-		$transient_version = WC_Cache_Helper::get_transient_version( 'woocommerce_reports' );
+		// Sync is taking place async, hence we bound to the transient version.
+		// With high orders placement rate, this caching here will be short-lived (suboptimal for BFCM/Christmas and busy stores in general).
+		$cache_version = WC_Cache_Helper::get_transient_version( 'woocommerce_reports' );
+	} elseif ( '' === $customer_email && $user_id ) {
+		// Optimized: for specific customers rely on orders count (customer meta under the hood with in-memory caching in the way to it).
+		// Best-case scenario for caching here, as it only depends on the customer orders placement.
+		$cache_version = wc_get_customer_order_count( $user_id );
```

Comment:
The counter is a user meta. At this point, the user object is fully populated and present somewhere in memory - hence we can skip transients here altogether.